File: src/main/java/audioWfc/wfc/OptionsPerCell.java
package audioWfc.wfc;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class OptionsPerCell<T> {
    private Map<Integer, Set<T>> cells;
    private Set<T> allOptions;
    private int size;

    public OptionsPerCell(Set<T> allOptions){
        this.cells = new HashMap<>();
        this.allOptions = allOptions;
    }

    public void reset(){
        this.cells = new HashMap<>();
    }

    public void setOptions(int pos, Set<T> options){
        cells.put(pos, options);
    }

    public void setValue(int pos, T value){
        cells.put(pos, Set.of(value));
    }

    public Set<T> get(int pos){
        if(!cells.containsKey(pos)) return allOptions;
        return cells.get(pos);
    }

    public Set<T> getAllOptions() {
        return allOptions;
    }

    public OptionsPerCell<T> union(OptionsPerCell<T> other){
        Set<T> newAllOptions = new HashSet<>();
        newAllOptions.addAll(this.allOptions);
        newAllOptions.addAll(other.allOptions);
        OptionsPerCell<T> out = new OptionsPerCell<>(newAllOptions);

        for(Map.Entry<Integer, Set<T>> entry : this.cells.entrySet()){
            out.setOptions(entry.getKey(), entry.getValue());
        }

        for(Map.Entry<Integer, Set<T>> entry : other.cells.entrySet()){
            out.setOptions(entry.getKey(), entry.getValue());
        }

        return out;
    }
}

File: vite-project/src/wfc/OptionsPerCell.ts
export class OptionsPerCell<T> {
	cells : Map<number, T[]>
	allOptions : T[]

	constructor(allOptions : T[], map? : Map<number, T[]>) {
		this.cells = map ?? new Map<number, T[]>()
		this.allOptions = allOptions
	}

	public reset(): void {
		this.cells.clear()
	}

	public setOptions(position : number, options : T[]): void {
		this.cells.set(position, options)
	}

	public setValue(position : number, value : T): void {
		this.cells.set(position, [value])
	}

	public getOptions(position : number): T[] {
		return this.cells.get(position) || this.allOptions
	}

	public getAllOptions(): T[] {
		return this.allOptions
	}

	public union(other: OptionsPerCell<T>): OptionsPerCell<T> {
		const newAllOptions = [...this.allOptions, ...other.allOptions]
    
		const newOptionsPerCell = new OptionsPerCell<T>(newAllOptions)

		for(const [position, options] of this.cells) {
			newOptionsPerCell.setOptions(position, options)
		}
		for(const [position, options] of other.cells) {
			newOptionsPerCell.setOptions(position, options)
		}

		return newOptionsPerCell
	}


}